Add default implementations for FixedWidthInteger.dividingFullWidth

stephentyrone opened the pull request:
These are provided for FixedWidthInteger & UnsignedInteger (the base implementation, following Knuth's Algorithm D) and SignedInteger (converting to magnitudes and calling the former). Previously no default implementations were available, requiring every type to implement these operations.

These defaults will not be optimal for large fixed-width integers, so types vending Int512 or similar integers should still provide their own implementations, but they are unconditionally available as a fallback, which simplifies the process of writing such types, and work well enough as a fallback for modest fixed-width integer types like Int64 or 32b or smaller platforms or Int128 on 64b platforms.

Additionally rework the concrete implementations to guarantee that we always trap when the quotient is not representable, and to improve performance for 64b integers on `arm64_32`

Fixes #54509.

stephentyrone committed:
Add default implementations for FixedWidthInteger.dividingFullWidth

These are provided for FixedWidthInteger & UnsignedInteger (the base implementation, following Knuth's Algorithm D) and SignedInteger (converting to magnitudes and calling the former). Previously no default implementations were available, requiring every type to implement these operations. These defaults will not be optimal for large fixed-width integers, so types vending Int512 or similar types should still provide their own implementations, but they are unconditionally available as a fallback, which simplifies the process of writing such types, and work well enough as a fallback for modest fixed-width integer types like Int64 or 32b or smaller platforms or Int128 on 64b platforms.

stephentyrone commented:
@swift-ci test

stephentyrone committed:
Add missing @_aEIC attribute

stephentyrone commented:
@swift-ci test

phausler commented on stdlib/public/core/Integers.swift:
```diff
@@ -3491,9 +3502,124 @@ extension UnsignedInteger where Self: FixedWidthInteger {
+      // the quotient digit q, satisfying q ≤ q̂ ≤ q + 2. We adjust it
+      // downward as needed until we have the correct q.
+      while q̂ > mask || q̂ &* vl > (r̂ &<< n_2 | low) {
+        q̂ &-= 1
```
❤️ the use of circumflex

natecook1000 replied:
omg

phausler commented on stdlib/public/core/Integers.swift:
```diff
@@ -3491,9 +3502,124 @@ extension UnsignedInteger where Self: FixedWidthInteger {
+    // later on (see Knuth, Algorithm D), but when the divisor is only two
+    // half-words (as here), that can never happen, because we use the full
+    // divisor in the check for the while loop.
+    func generateHalfDigit(high: Self, low: Self) -> Self {
```
I presume this as an inner function is 100% specialized and inlined

stephentyrone replied:
So long as the outer function is specialized, yes.

phausler commented on stdlib/public/core/Integers.swift:
```diff
@@ -3491,9 +3502,124 @@ extension UnsignedInteger where Self: FixedWidthInteger {
+    // on Self as a word / halfword -> halfword division get one halfword
+    // digit of the quotient at a time.
+    let n_2 = Self.bitWidth/2
+    let mask = Self(1) &<< n_2 &- 1
```
depending on the register spill here; would it be worthwhile to optimize by adding additional variables like something to represent the masked value of `ul | mask` etc?

stephentyrone replied:
It ought not to matter, so I'll hold off until we see a case where it does.

phausler reviewed (approved):
The algorithm looks spot on to me; I presume we already have tests that cover this?

stephentyrone commented:
Most normal standard library tests won't exercise it because standard library integer types specialize the operation. I wrote a bunch of standalone tests for it, which I'll see about rolling into the test suite.

stephentyrone committed:
Use new default dividingFullWidth for 64b on 32b targets.

glessard reviewed (approved):
Nice explanations along the way!

stephentyrone committed:
Rework concrete dividingFullWidth implementation

- make these trap when quotient is not representable
- use hardware divide for 64b integers on arm64_32
- defer to new default implementation
- add some test coverage

stephentyrone commented:
@swift-ci test

stephentyrone commented:
@swift-ci benchmark

stephentyrone commented:
(We don't expect benchmarks to move much from this, but it's good to make sure.)

slavapestov reviewed (approved):
I love it!

stephentyrone commented:
As expected, benchmarks are just noisy tests, nothing related to division. Importantly, no noticeable code size changes at all.
